File: JavaScript/src/Internals/Functions/MathSum/Function_SUMPRODUCT.js
import { Function_N } from '../Function_N.js';
import { Operand } from '../../../Operand.js';

class Function_SUMPRODUCT extends Function_N {
    get Name() {
        return "SUMPRODUCT";
    }

    constructor(funcs) {
        super(funcs);
    }

    evaluate(engine, tempParameter) {
        if (this.z.length < 2) return this.parameterError(1);

        let arrays = [];
        for (let i = 0; i < this.z.length; i++) {
            let arg = this.getArray(engine, tempParameter, i);
            if (arg.IsError) return arg;
            let list = [];
            for (let item of arg.ArrayValue) {
                if (item.IsNumber) {
                    list.push(item.NumberValue);
                }
            }
            arrays.push(list);
        }

        let minLength = arrays[0].length;
        for (let i = 1; i < arrays.length; i++) {
            if (arrays[i].length < minLength) {
                minLength = arrays[i].length;
            }
        }

        if (minLength == 0) {
            return Operand.Create(0);
        }

        let result = 0;
        for (let i = 0; i < minLength; i++) {
            let product = 1;
            for (let j = 0; j < arrays.length; j++) {
                product *= arrays[j][i];
            }
            result += product;
        }

        return Operand.Create(result);
    }
}

export { Function_SUMPRODUCT };

File: java/src/main/java/toolgood/algorithm/internals/functions/mathsum/Function_SUMPRODUCT.java
package toolgood.algorithm.internals.functions.mathsum;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_N;
import toolgood.algorithm.internals.functions.NoneEngine;

public final class Function_SUMPRODUCT extends Function_N {
    public Function_SUMPRODUCT(FunctionBase[] funcs) {
        super(funcs);
    }

    @Override
    public String Name() {
        return "SUMPRODUCT";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, java.util.function.BiFunction<AlgorithmEngine, String, Operand> tempParameter) throws Exception {
        if (funcs.length < 2) return ParameterError(1);

        List<List<BigDecimal>> arrays = new ArrayList<>();
        for (int i = 0; i < funcs.length; i++) {
            Operand arg = GetArray(engine, tempParameter, i);
            if (arg.IsErrorOrNone()) return arg;
            List<BigDecimal> list = new ArrayList<>();
            for (Operand item : arg.ArrayValue()) {
                if (item.IsNumber()) {
                    list.add(item.NumberValue());
                }
            }
            arrays.add(list);
        }

        int minLength = arrays.get(0).size();
        for (int i = 1; i < arrays.size(); i++) {
            if (arrays.get(i).size() < minLength) {
                minLength = arrays.get(i).size();
            }
        }

        if (minLength == 0) {
            return Operand.Zero;
        }

        BigDecimal result = BigDecimal.ZERO;
        for (int i = 0; i < minLength; i++) {
            BigDecimal product = BigDecimal.ONE;
            for (int j = 0; j < arrays.size(); j++) {
                product = product.multiply(arrays.get(j).get(i));
            }
            result = result.add(product);
        }

        return Operand.Create(result);
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        for (int i = 0; i < funcs.length; i++) {
            funcs[i].GetParameterTypes(noneEngine, result, OperandType.ARRAY);
        }
    }
}
